lib/Transforms/ShapeInference/ShapeInference.cpp

```diff
@@ -0,0 +1,187 @@
+
+    auto walkResult = funcOp.getBody().walk([&](Operation *op) {
+      // We focus on operations with return type inference,
+      // as these will likely complain if their operands' types change
+      // but their return type hasn't been updated.
+      // Technically, other ops could also cause issues,
+      // For example, an op could allow only dynamically shaped tensors,
+      // which means it would fail to verify after shape inference,
+      // but failing on those is probably the right thing to do for the compiler
+      if (auto inferOp = dyn_cast<InferTypeOpInterface>(op)) {
+        SmallVector<Type> inferredReturnTypes;
+        if (failed(inferOp.inferReturnTypes(
+                rewriter.getContext(), op->getLoc(), op->getOperands(),
+                op->getAttrDictionary(), op->getPropertiesStorage(),
+                op->getRegions(), inferredReturnTypes))) {
+          emitError(op->getLoc(), "Failed to infer return types.");
+          return WalkResult::interrupt();
+        }
+
+        assert(
+            inferredReturnTypes.size() == op->getNumResults() &&
+            "Number of inferred return types must match the number of results");
+
+        for (size_t i = 0; i < inferredReturnTypes.size(); ++i) {
+          if (inferredReturnTypes[i] == op->getResult(i).getType()) continue;
+          op->getResult(i).setType(inferredReturnTypes[i]);
+        }
+      }
+      return WalkResult::advance();
+    });
+    if (walkResult.wasInterrupted()) {
+      rewriter.cancelOpModification(funcOp);
+      return failure();
+    }
+
+    /////////////////////////////////////////////////////////////////////
+    //// Step 4: Match the return type(s) in the signature           ////
+    /////////////////////////////////////////////////////////////////////
+
+    // Get Function Signature return types
+    auto returnTypes = funcOp.getFunctionType().getResults();
+
+    // Get terminator (func.return) types
+    auto *terminator = funcOp.getBody().getBlocks().front().getTerminator();
+    assert(terminator && "missing required terminator in func op.");
+    auto terminatorTypes = terminator->getOperandTypes();
+
+    assert(returnTypes.size() == terminatorTypes.size() &&
+           "Expected same number of results in signature and terminator.");
+
+    funcOp.setFunctionType(rewriter.getFunctionType(
+        funcOp.getFunctionType().getInputs(), terminatorTypes));
+
+    // Signal that we have successfully modified the op
+    rewriter.finalizeOpModification(funcOp);
+    return success();
+  }
+};
+
+struct ShapeInference : impl::ShapeInferenceBase<ShapeInference> {
+  using ShapeInferenceBase::ShapeInferenceBase;
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+    // FIXME: This probably shouldn't even be a pattern anymore...
+    patterns.add<InferShape>(context);
+    (void)applyPatternsGreedily(getOperation(), std::move(patterns));
```

Comment:
```suggestion
    patterns.add<InferShape>(context);
    walkAndApplyPatterns(getOperation(), std::move(patterns));
```

IMO it's fine to be a pattern, but no need for greedy rewrite engine